grand_challenge_forge/partials/example-evaluation-method/example-evaluation-method{{cookiecutter._}}/helpers.py

```diff
@@ -0,0 +1,164 @@
+
+    Parameters
+    ----------
+    fn : function
+        Function to execute that will process each prediction
+
+    predictions : list
+        List of predictions.
+
+    Returns
+    -------
+    A list of results
+    """
+    with Manager() as manager:
+        results = manager.list()
+        errors = manager.list()
+
+        pool_worker = _start_pool_worker(
+            fn=fn,
+            predictions=predictions,
+            max_workers=get_max_workers(),
+            results=results,
+            errors=errors,
+        )
+        try:
+            pool_worker.join()
+        finally:
+            pool_worker.terminate()
+
+        for prediction, e in errors:
+            raise PredictionProcessingError(
+                prediction=prediction,
+                error=e,
+            ) from e
+
+        return list(results)
+
+
+def _start_pool_worker(fn, predictions, max_workers, results, errors):
+    process = Process(
+        target=_pool_worker,
+        name="PredictionProcessing",
+        kwargs=dict(
+            fn=fn,
+            predictions=predictions,
+            max_workers=max_workers,
+            results=results,
+            errors=errors,
+        ),
+    )
+    process.start()
+
+    return process
+
+
+def _pool_worker(*, fn, predictions, max_workers, results, errors):
+    terminating_child_processes = False
+
+    with ProcessPoolExecutor(max_workers=max_workers) as executor:
+        try:
+
+            def handle_error(error, prediction="Unknown"):
+                executor.shutdown(wait=False, cancel_futures=True)
+                errors.append((prediction, error))
+
+                nonlocal terminating_child_processes
+                terminating_child_processes = True
+                _terminate_child_processes()
+
+            def sigchld_handler(*_, **__):
+                if not terminating_child_processes:
+                    handle_error(
+                        RuntimeError(
+                            "Child process was terminated unexpectedly"
+                        )
+                    )
+
+            # Register the SIGCHLD handler
+            signal.signal(signal.SIGCHLD, sigchld_handler)
+
+            # Submit the processing tasks of the predictions
+            futures = [
+                executor.submit(fn, prediction) for prediction in predictions
+            ]
+            future_to_predictions = {
+                future: item
+                for future, item in zip(futures, predictions, strict=True)
+            }
+
+            for future in as_completed(future_to_predictions):
+                try:
+                    result = future.result()
+                    results.append(result)
+                except Exception as e:
+                    handle_error(e, prediction=future_to_predictions[future])
+        finally:
+            terminating_child_processes = True
+            _terminate_child_processes()
```

Comment:
Sorry for the back and forth here - and especially delay with the review because of the conference. But from my experiments, we can likely just remove the entire safe-guards here again. Again, I have definitely seen child-processes become zombies and clog up systems but I was not able to reproduce this again. Information might just be outdated as my experience with this is already 5+ years old... So if you agree, I can help you rewind the entire SIGCHLD handling again?

Reply:
It is still an issue in general - but maybe not with the handling added here? You're not doing something like setting `--init` by default to docker run, or by having `init: true` in your docker compose file?

Reply:
AFAIK the problem in this case is that ANY child of the spawned processes might be killed by OOM and hence would leave a parent waiting for it to come home.

The tests, I think, are simple but solid examples. I vote to keep the safeguards since we've stretched the upper limit of runtime of containers up to 28 days...

Reply:
Sounds good to me.

Reply:
Also fine by me!

Reply:
Just checking: are you going to review the content still or are you generally ok with it? If the first, no worries take your time. If the latter, could you approve it formally?

Reply:
Oh sorry. I somehow missed your last post and also forgot to approve this in general, you are right!